stablehlo/compatibility/transforms/CompatibilityPasses.cpp

```diff
@@ -0,0 +1,621 @@
+    auto failOrVersion = VersionNumber::get(targetVersion);
+    if (failed(failOrVersion)) {
+      return emitError(
+          getOperation()->getLoc(),
+          "invalid target version number argument " + targetVersion);
+    }
+    VersionNumber targetVersionNumber = *failOrVersion;
+    if (targetVersionNumber < VersionNumber::getMinimumSupported()) {
+      return emitError(getOperation()->getLoc())
+             << "target version " << targetVersion
+             << " is less than minimum supported";
+    }
+    if (VersionNumber::getCurrent() < targetVersionNumber) {
+      return emitError(getOperation()->getLoc())
+             << "target version " << targetVersion
+             << " is greater than current version";
+    }
+    return targetVersionNumber;
+  }
+
+  void runOnOperation() override {
+    ConversionTarget target(getContext());
+
+    // Validate version number
+    auto failOrVersion = validateTargetVersion(targetVersion);
+    if (failed(failOrVersion)) {
+      return signalPassFailure();
+    }
+    VersionNumber targetVersionNumber = *failOrVersion;
+
+    // An op is legal if the target version is in the ops `[min, max]`
+    // supported version range.
+    // Example:
+    //   CustomCallv1 0.0.0 -> 0.0.x
+    //   CustomCallv2 0.1.0 -> 0.4.x
+    //   CustomCallv3 0.5.0 -> Curr
+    // Target Curr (0.5.0):
+    //   v3 legal    { Curr  in [0.5, Curr] }
+    //   v2 illegal  { Curr !in [0.1, 0.4] }
+    //   v1 illegal  { Curr !in [0.0, 0.0] }
+    // Target 0.4.0:
+    //   v3 illegal { 0.4 !in [0.5, Curr] }
+    //   v2 legal   { 0.4  in [0.1, 0.4] }
+    //   v1 illegal { 0.4 !in [0.0, 0.0] }
+    // Target 0.0.0:
+    //   v3 illegal { 0.0 !in [0.5, Curr] }
+    //   v2 illegal { 0.1 !in [0.1, 0.4] }
+    //   v1 legal   { 0.0  in [0.0, 0.1] }
+    target.addDynamicallyLegalDialect<VersionedhloDialect>([&targetVersionNumber](
+                                                               Operation* op) {
+      if (auto interface = dyn_cast<VersionInterface>(op)) {
+        return (interface.getMinVersion() <= targetVersionNumber &&
+                targetVersionNumber <= interface.getMaxVersion());
+      }
+      return false;
+    });
+
+    versionedhlo::VersionedTypeConverterBase converter;
+    RewritePatternSet patterns(&getContext());
+    versionedhlo::populateVersionedhloToVersionPatterns(&patterns, &converter,
+                                                        &getContext());
+
+    // Conversion from VersionedHLO to StableHLO should never fail.
+    if (failed(applyPartialConversion(getOperation(), target,
+                                      std::move(patterns)))) {
+      return signalPassFailure();
+    }
+  }
+};
+
+////////////////////////////////////////////
+/// Upgrade and Downgrade Infrastructure ///
+////////////////////////////////////////////
+
+LogicalResult emitToVersionError(Operation* op, llvm::StringRef message) {
+  return op->emitError("failed to downgrade ")
```

Comment:
Is the concern that this applies to `ToVersion` in both directions? 

This is only used for downgrades since upgrades should never fail. Renamed the function to `emitDowngradeError`.